File: tools/sdk-ai-bots/azure-sdk-qa-bot-knowledge-sync/src/services/AppSecret.ts
import { AzureCliCredential, ChainedTokenCredential, ManagedIdentityCredential, WorkloadIdentityCredential} from '@azure/identity';
import { SecretClient } from '@azure/keyvault-secrets';

/**
 * Initialize secrets by loading them from Azure Key Vault and setting to environment variables
 */
export async function initSecrets(): Promise<void> {
    // Get the Key Vault endpoint from environment variable
    const keyVaultEndpoint = process.env.KEYVAULT_ENDPOINT;
    
    if (!keyVaultEndpoint) {
        throw new Error('KEYVAULT_ENDPOINT environment variable is required');
    }

    try {
        console.log('Loading secrets from Azure Key Vault...');
        
        // Create a credential
        const credential = new ChainedTokenCredential(
            new ManagedIdentityCredential(),
            new AzureCliCredential(),
            new WorkloadIdentityCredential()
        );

        // Establish a connection to the Key Vault client
        const client = new SecretClient(keyVaultEndpoint, credential);

        // Get AI Search API Key
        const aiSearchSecretResponse = await client.getSecret('AI-SEARCH-APIKEY');
        if (!aiSearchSecretResponse.value) {
            throw new Error('Failed to get AI-SEARCH-APIKEY secret value');
        }

        process.env.AI_SEARCH_API_KEY = aiSearchSecretResponse.value;
        console.log('Set AI_SEARCH_API_KEY from Key Vault');

        // Get AOAI Chat Completions API Key
        const aoaiSecretResponse = await client.getSecret('AOAI-CHAT-COMPLETIONS-API-KEY');
        if (!aoaiSecretResponse.value) {
            throw new Error('Failed to get AOAI-CHAT-COMPLETIONS-API-KEY secret value');
        }

        process.env.AOAI_CHAT_COMPLETIONS_API_KEY = aoaiSecretResponse.value;
        console.log('Set AOAI_CHAT_COMPLETIONS_API_KEY from Key Vault');

        // Get SSH private key
        const sshPrivateKeySecret = await client.getSecret('SSH-PRIVATE-KEY');
        if (!sshPrivateKeySecret.value) {
            throw new Error('Failed to get SSH-PRIVATE-KEY secret value');
        }
        
        process.env.SSH_PRIVATE_KEY = sshPrivateKeySecret.value;

        console.log('Successfully loaded secrets from Azure Key Vault');

    } catch (error) {
        console.error('Failed to load secrets from Azure Key Vault:', error);
        throw error;
    }
}

File: tools/sdk-ai-bots/azure-sdk-qa-bot-knowledge-graph-sync/azure_sdk_qa_bot_knowledge_graph_sync/services/app_secret.py
"""Azure Key Vault secret loader.

Fetches required secrets from Key Vault and sets them as environment variables.
"""

from __future__ import annotations

import logging
import os

logger = logging.getLogger(__name__)


async def init_secrets() -> None:
    """Load secrets from Azure Key Vault into environment variables."""
    endpoint = os.environ.get("KEYVAULT_ENDPOINT")
    if not endpoint:
        logger.warning("KEYVAULT_ENDPOINT not set; skipping Key Vault secrets")
        return

    from azure.identity.aio import (
        AzureCliCredential,
        ChainedTokenCredential,
        ManagedIdentityCredential,
        WorkloadIdentityCredential,
    )
    from azure.keyvault.secrets.aio import SecretClient

    logger.info("Loading secrets from Azure Key Vault...")

    credentials = []
    for cls in (WorkloadIdentityCredential, AzureCliCredential, ManagedIdentityCredential):
        try:
            credentials.append(cls())
        except Exception as e:  # noqa: BLE001
            logger.debug("Skipping %s: %s", cls.__name__, e)
    if not credentials:
        raise RuntimeError("No Azure credentials available for Key Vault")
    credential = ChainedTokenCredential(*credentials)

    client = SecretClient(vault_url=endpoint, credential=credential)

    try:
        # AI Search API Key
        secret = await client.get_secret("AI-SEARCH-APIKEY")
        if secret.value:
            os.environ["AI_SEARCH_API_KEY"] = secret.value
            logger.info("Set AI_SEARCH_API_KEY from Key Vault")

        # Azure OpenAI API Key
        secret = await client.get_secret("AOAI-CHAT-COMPLETIONS-API-KEY")
        if secret.value:
            os.environ["AOAI_CHAT_COMPLETIONS_API_KEY"] = secret.value
            logger.info("Set AOAI_CHAT_COMPLETIONS_API_KEY from Key Vault")

        # SSH private key
        secret = await client.get_secret("SSH-PRIVATE-KEY")
        if secret.value:
            os.environ["SSH_PRIVATE_KEY"] = secret.value
            logger.info("Set SSH_PRIVATE_KEY from Key Vault")

    finally:
        await client.close()
        await credential.close()

    logger.info("Successfully loaded secrets from Azure Key Vault")
